homeassistant/components/opentherm_gw/entity.py

```diff
@@ -74,3 +62,63 @@ def receive_report(self, status: dict[str, dict]) -> None:
         """Handle status updates from the component."""
         # Must be implemented at the platform level.
         raise NotImplementedError
+
+
+class OpenThermBoilerDeviceEntity(OpenThermBaseEntity):
+    """Represent an OpenTherm Boiler entity."""
+
+    _data_source = OpenThermDataSource.BOILER
+
+    def __init__(
+        self,
+        gw_hub: OpenThermGatewayHub,
+        description: OpenThermEntityDescription,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(gw_hub, description)
+        self._attr_unique_id = (
+            f"{gw_hub.hub_id}-{DEVICE_IDENT_BOILER}-{description.key}"
+        )
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{gw_hub.hub_id}-{DEVICE_IDENT_BOILER}")},
+        )
+
+
+class OpenThermGatewayDeviceEntity(OpenThermBaseEntity):
+    """Represent an OpenTherm Gateway entity."""
+
+    _data_source = OpenThermDataSource.GATEWAY
+
+    def __init__(
+        self,
+        gw_hub: OpenThermGatewayHub,
+        description: OpenThermEntityDescription,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(gw_hub, description)
```

Comment:
I will rethink the strategy here.

Reply:
> I think you can move the full device creation you now do in `__init__.py` to here ...

We still need to reference the devices in `__init__.py` to handle updates to the device info. It doesn't make sense to do that at the entity level. This is why I chose to create them from there.

Reply:
device_info is only read once when the entities are added. The rest goes via the device registry.

Reply:
I am well aware of that. What I mean is that we store a reference to the device entry here: https://github.com/mvn23/home-assistant/blob/61ed9a3e0f79d5726529264239da6c906b5fabfd/homeassistant/components/opentherm_gw/__init__.py#L486-L495
so that we can use it to update the device info (manufacturer, model, etc) in `handle_report` here: https://github.com/mvn23/home-assistant/blob/61ed9a3e0f79d5726529264239da6c906b5fabfd/homeassistant/components/opentherm_gw/__init__.py#L504-L528

This information is not immediately available on startup, but only comes in later on. This is why we have to update the devices upon receiving a report from the gateway and not while adding the device. If we perform this update at the entity level, `async_update_device` will be called from every single entity which does not seem desirable to me.

Technically it would be sufficient to store only the device id, but regardless of that it still makes sense to `get_or_create` the devices in `__init__.py`.

Reply:
In that case, agreed